### DISCUSSION TITLE
[Python] skywalking [MainThread] [WARNING] fork() not currently supported with grpc protocol

### DISCUSSION THREAD

**Discussion by rysinal** [apache/skywalking#8518] [Q&A] [answered]
I refer to the [Python Agent documentation](https://skywalking.apache.org/docs/skywalking-python/latest/en/setup/cli/#skywalking-python-agent-command-line-interfacecli) to start my program via the start command `sw-python -d run gunicorn myapp.wsgi` , When I access the api I get an error:

```
skywalking [MainThread] [WARNING] fork() not currently supported with grpc protocol
skywalking [ReportThread] [ERROR] <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INTERNAL
        details = "Received RST_STREAM with error code 5"
        debug_error_string = "{"created":"@1644385809.129550000","description":"Error received from peer ipv4:172.23.76.178:11800","file":"src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Received RST_STREAM with error code 5","grpc_status":13}"
>
skywalking [SendProfileSnapShotThread] [ERROR] <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INTERNAL
        details = "Received RST_STREAM with error code 5"
        debug_error_string = "{"created":"@1644385809.837323000","description":"Error received from peer ipv4:172.23.76.178:11800","file":"src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Received RST_STREAM with error code 5","grpc_status":13}"
```

Instead, I was able to succeed with the following startup command

```
sw-python -d run python3 manage.py runserver
```

My sample code like this:

```
from django.http import HttpResponse

def index(request):
    return HttpResponse("Hello, world. You're at the sky index.")
```

my env:

Python 3.8.9
SkyWalking 8.8.1
gunicorn (version 20.1.0)

is it possible that gunicorn's startup method requires special settings to support skywalking? Please tell me how to use the gunicorn command to avoid this error，Thank you so much

**Comment by kezhenxu94** [comment, marked as answer]
As the error message says, `[MainThread] [WARNING] fork() not currently supported with grpc protocol`, your application is somewhat using `fork` process that is not supported in gRPC protocol, you need to switch to HTTP protocol.

**Reply by rysinal** [reply]
tks, This error has been solved when i set the env of SW_AGENT_PROTOCOL to 'http', The backend OAP server address port modify to 12800. 
but i got another err:

```
skywalking-loader [MainThread] [DEBUG] Original sitecustomize module not found, skipping.
skywalking-loader [MainThread] [DEBUG] SkyWalking Python Agent starting, loader finished.
skywalking-loader [MainThread] [ERROR] SkyWalking Python Agent failed to start, please inspect your package installation
[2022-02-10 14:56:47 +0800] [76293] [INFO] Starting gunicorn 20.1.0
[2022-02-10 14:56:47 +0800] [76293] [INFO] Listening at: http://127.0.0.1:8000 (76293)
[2022-02-10 14:56:47 +0800] [76293] [INFO] Using worker: sync
[2022-02-10 14:56:47 +0800] [76295] [INFO] Booting worker with pid: 76295
```

**Reply by rysinal** [reply]
@kezhenxu94 Can you help me if you have time? tks

**Reply by kezhenxu94** [reply]
> @kezhenxu94 Can you help me if you have time? tks

I didn't get notified when you edited an original comment, so please create a new comment if you have more to add, after you've replied for some time.

Looks like the exception of `agent.start()` is not printed, can you tried to add that exception into the log and see what's the exception? 

https://github.com/apache/skywalking-python/blob/9b0f83d74f37650dcff6082ff2d738dd2aee49d9/skywalking/bootstrap/loader/sitecustomize.py#L147-L149

If you like you can open a pull request to add this, after exposing the exception we can investigate what's going on

**Reply by Superskyyy** [reply]
Yes, please change the line `logger.error()` to `logger.exception()`, then it should do the trick to output tracebacks.

**Reply by Superskyyy** [reply]
https://github.com/apache/skywalking-python/pull/183 @rysinal Please try again from the latest master after this one merges, so we can together find out the root cause.

**Reply by rysinal** [reply]
> [apache/skywalking-python#183](https://github.com/apache/skywalking-python/pull/183) @rysinal Please try again from the latest master after this one merges, so we can together find out the root cause.

Maybe because the computer rebooted, I can't reproduce it now, if I do I'll post the log to this page again, thanks for your answer
